### Applying the Machine learning algorithm (LInear regression) on AirQuality dataset available on UCI http://archive.ics.uci.edu/ml/datasets/Air+Quality

In [1]:
# Importing basic libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Reading the dataset
dataset=pd.read_excel(r'C:\Users\Hemant Raturi\Downloads\DA & ML\ML AirQualityUCI.xlsx')
dataset.head(50)

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)   C6H6(GT)  \
0  2004-03-10  18:00:00     2.6      1360.00       150  11.881723   
1  2004-03-10  19:00:00     2.0      1292.25       112   9.397165   
2  2004-03-10  20:00:00     2.2      1402.00        88   8.997817   
3  2004-03-10  21:00:00     2.2      1375.50        80   9.228796   
4  2004-03-10  22:00:00     1.6      1272.25        51   6.518224   
5  2004-03-10  23:00:00     1.2      1197.00        38   4.741012   
6  2004-03-11  00:00:00     1.2      1185.00        31   3.624399   
7  2004-03-11  01:00:00     1.0      1136.25        31   3.326677   
8  2004-03-11  02:00:00     0.9      1094.00        24   2.339416   
9  2004-03-11  03:00:00     0.6      1009.75        19   1.696658   
10 2004-03-11  04:00:00  -200.0      1011.00        14   1.293620   
11 2004-03-11  05:00:00     0.7      1066.00         8   1.133431   
12 2004-03-11  06:00:00     0.7      1051.75        16   1.603768   
13 2004-03-11  07:00:00     1.1      1144.00        29   3.243618   
14 2004-03-11  08:00:00     2.0      1333.25        64   8.013773   
15 2004-03-11  09:00:00     2.2      1351.00        87   9.540643   
16 2004-03-11  10:00:00     1.7      1233.25        77   6.335782   
17 2004-03-11  11:00:00     1.5      1178.75        43   4.971584   
18 2004-03-11  12:00:00     1.6      1236.00        61   5.216919   
19 2004-03-11  13:00:00     1.9      1285.50        63   7.269933   
20 2004-03-11  14:00:00     2.9      1371.00       164  11.539007   
21 2004-03-11  15:00:00     2.2      1310.00        79   8.826223   
22 2004-03-11  16:00:00     2.2      1291.75        95   8.301413   
23 2004-03-11  17:00:00     2.9      1383.00       150  11.151581   
24 2004-03-11  18:00:00     4.8      1580.75       307  20.799217   
25 2004-03-11  19:00:00     6.9      1775.50       461  27.359807   
26 2004-03-11  20:00:00     6.1      1640.00       401  24.017757   
27 2004-03-11  21:00:00     3.9      1312.75       197  12.779368   
28 2004-03-11  22:00:00     1.5       964.50        61   4.707072   
29 2004-03-11  23:00:00     1.0       912.75        26   2.645722   
30 2004-03-12  00:00:00     1.7      1080.25        55   5.854802   
31 2004-03-12  01:00:00     1.9      1043.75        53   6.374298   
32 2004-03-12  02:00:00     1.4       987.75        40   4.132342   
33 2004-03-12  03:00:00     0.8       888.75        21   1.869445   
34 2004-03-12  04:00:00  -200.0       831.00        10   1.068293   
35 2004-03-12  05:00:00     0.6       847.25         7   1.022415   
36 2004-03-12  06:00:00     0.8       927.00        17   1.830431   
37 2004-03-12  07:00:00     1.4      1090.50        33   4.359341   
38 2004-03-12  08:00:00     4.4      1587.00       202  17.865587   
39 2004-03-12  09:00:00  -200.0      1544.50      -200  22.074162   
40 2004-03-12  10:00:00     3.1      1350.25       208  14.027011   
41 2004-03-12  11:00:00     2.7      1262.75       166  11.645647   
42 2004-03-12  12:00:00     2.1      1206.25       114  10.224662   
43 2004-03-12  13:00:00     2.5      1251.50       140  11.039936   
44 2004-03-12  14:00:00     2.7      1287.00       169  12.816446   
45 2004-03-12  15:00:00     2.9      1352.75       185  14.173851   
46 2004-03-12  16:00:00     2.8      1309.00       165  12.690568   
47 2004-03-12  17:00:00     2.4      1274.00       133  11.738405   
48 2004-03-12  18:00:00     3.9      1509.50       233  19.290975   
49 2004-03-12  19:00:00     3.7      1525.25       242  18.226178   

    PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1045.50    166.0       1056.25    113.0       1692.00      1267.50   
1          954.75    103.0       1173.75     92.0       1558.75       972.25   
2          939.25    131.0       1140.00    114.0       1554.50      1074.00   
3          948.25    172.0       1092.00    122.0       1583.75      1203.25   
4          835.50    131.0       1205.00    116.0       1490.00      1110.00   
5        

Since we have three values(Temprature, Relative humidity, Absolute humidity) to predict we will take three different prediction vector y,y1,y2

In [3]:
# Selecting feature matrix(X) and prediction vector(y)

# selecting columns for X
X=dataset.iloc[:,2:12].values
# selecting columns for y Temprature
y=dataset.iloc[:,[12]].values 
# selecting columns for y Relative humidity
y1=dataset.iloc[:,[13]].values 
# selecting columns for y Absolute humidity
y2=dataset.iloc[:,[14]].values 

### Preprocessing the data

Missing values are tagged with -200 value

In [4]:
# Replacing Misssing values
from sklearn.impute import SimpleImputer
sim = SimpleImputer(missing_values=-200, strategy='mean')
X = sim.fit_transform(X)
X

array([[   2.6       , 1360.        ,  150.        , ...,  113.        ,
        1692.        , 1267.5       ],
       [   2.        , 1292.25      ,  112.        , ...,   92.        ,
        1558.75      ,  972.25      ],
       [   2.2       , 1402.        ,   88.        , ...,  114.        ,
        1554.5       , 1074.        ],
       ...,
       [   2.4       , 1142.        ,  218.81181619, ...,  174.7       ,
        1240.75      , 1092.        ],
       [   2.1       , 1002.5       ,  218.81181619, ...,  155.7       ,
        1041.        ,  769.75      ],
       [   2.2       , 1070.75      ,  218.81181619, ...,  167.7       ,
        1128.5       ,  816.        ]])

In [5]:
# Removing rows where y is missing as their result would be misleading
missin_value_rows=np.where(y==-200) # Geeting rows with missing values
missin_value_rows[0]

array([ 524,  525,  526,  701,  702,  703,  704,  705,  706,  707,  708,
        709,  710,  711,  712,  713,  714,  715,  716,  717,  718,  719,
        720,  721,  722,  723,  724, 1825, 1826, 1827, 1828, 1829, 1830,
       1831, 1832, 1833, 1834, 1835, 1836, 1837, 1838, 2420, 2421, 2422,
       2423, 2424, 2425, 2426, 2427, 2428, 2429, 2430, 2431, 2432, 2433,
       2434, 2435, 2436, 2437, 2438, 2439, 2440, 2441, 2442, 2443, 2444,
       2445, 2446, 2447, 2448, 2449, 2450, 2451, 2452, 2453, 2454, 2455,
       2456, 2457, 3414, 3646, 3647, 3648, 3649, 4044, 4045, 4046, 4047,
       4048, 4049, 4050, 4051, 4052, 4053, 4054, 4055, 4056, 4057, 4058,
       4059, 4060, 4061, 4062, 4063, 4064, 4065, 4066, 4067, 4068, 4069,
       4070, 4071, 4072, 4073, 4074, 4075, 4076, 4077, 4078, 4079, 4080,
       4081, 4082, 4083, 4084, 4085, 4086, 4087, 4088, 4349, 4350, 4351,
       4352, 4353, 4354, 4355, 4356, 4357, 4358, 4360, 4361, 4362, 4363,
       4364, 4365, 4366, 4367, 4916, 6695, 6696, 66

In [6]:
# deleting from y
y=np.delete(y,missin_value_rows[0]) 
# deleting from y1
y1=np.delete(y1,missin_value_rows[0]) 
# deleting from y2
y2=np.delete(y2,missin_value_rows[0]) 
# Deleting from X also axis is given as it's a matrix
X=np.delete(X,missin_value_rows[0], axis=0) 

In [7]:
# Scaling the data using the standard Scaler

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

In [8]:
X

array([[ 3.59012596e-01,  1.19910236e+00, -1.06062221e+00, ...,
         1.62573001e-02,  6.81301380e-01,  6.14164662e-01],
       [-1.03167486e-01,  8.86994648e-01, -1.64650187e+00, ...,
        -4.70091866e-01,  2.96391897e-01, -1.26815453e-01],
       [ 5.08925418e-02,  1.39258610e+00, -2.01653113e+00, ...,
         3.94167842e-02,  2.84115234e-01,  1.28543486e-01],
       ...,
       [ 2.04952569e-01,  1.94829582e-01,  3.10520033e-04, ...,
         1.44519747e+00, -6.22191336e-01,  1.73717549e-01],
       [-2.61374718e-02, -4.47812859e-01,  3.10520033e-04, ...,
         1.00516727e+00, -1.19919448e+00, -6.35023661e-01],
       [ 5.08925418e-02, -1.33401773e-01,  3.10520033e-04, ...,
         1.28308108e+00, -9.46439662e-01, -5.18951416e-01]])

Applying Linear Regression model

In [9]:
# 3 Objects of LinearRegression for 3 differen valuess(Temprature, Relative humidity, Absolute humidity)

from sklearn.linear_model import LinearRegression
# LR for y
lin_reg=LinearRegression() 
# LR for y1
lin_reg1=LinearRegression() 
# LR for y2
lin_reg2=LinearRegression() 

In [10]:
# Fitting the model for Temprature
lin_reg.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [11]:
# Fitting the model for Relative humidity
lin_reg1.fit(X,y1)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [12]:
# Fitting the model for Absolute humidity
lin_reg2.fit(X,y2)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [13]:
# Checking the prediction score for Temprature
lin_reg.score(X,y)

0.6169310898748224

We get a score of 61% it is just fine

In [14]:
# Checking the prediction score for Relative humidity
lin_reg1.score(X,y1)

0.4452752611896203

We got a score of 44% which is bad

In [15]:
# Checking the prediction score for Absolute humidity
lin_reg2.score(X,y2)

0.7845955962007295

We got a score of 78% which is just fine.